### PR TITLE
Field require will respect dependsOn

### PR DESCRIPTION
**Pull request by snowkeeper**
Includes tests.

If a Field that has both `dependsOn` and `require: true` set, we first check for `evalDependsOn()`.
If `evalDependsOn` returns `true`, we will respect the `required` option. 
If `evalDependsOn` returns `false` we will ignore the `required` option.

### CODE REVIEW

**Comment by webteckie** [comment]
Cool stuff...left some nitpick comments :-)

File: fields/types/Type.js [hunk 1]
```diff
@@ -64,12 +64,25 @@ function Field(list, path, options) {
...
 		process.exit(1);
 	}
-
+	
+	// if dependsOn and required, set required to a function for validation
+	if (this.options.dependsOn && this.options.required === true) {
+		var opts = this.options;
+		this.options.required = function() {
+			// `this` refers to the validating document
+			debug('validate dependsOn required', evalDependsOn(opts.dependsOn, this.toObject()));
```
Comment (ericelliott) [inline fields/types/Type.js]: Is this needed?

**Comment by JedWatson** [comment]
@webteckie I'm working on the linter now actually. Since they stopped enabling rules by default in ESLint, we've actually had some massive blind spots. Will be much more robust soon :)

File: fields/types/Type.js (hunk 1 above)
Comment (snowkeeper) [inline fields/types/Type.js]: Without it working in the Admin UI would produce a different result that working on a List directly.

**Comment by JedWatson** [comment]
This looks like a really good enhancement, @snowkeeper, except for one concern -

I am pretty sure Mongoose will validate `required` paths regardless of whether Keystone pre-validates them or not, so this would still actually fail but with mongoose error messages instead of Keystone's higher-level ones. May not be true of all field types (e.g. I don't think we specify `required` on complex field sub-paths, which may actually be something we should do...)

This needs to be looked into further; another option may be to _exclude_ the `required` option from the Mongoose schema if `dependsOn` is also set, and live with the limitation that validation will then only be done through Keystone's APIs (e.g. `updateHandler`)

**Comment by snowkeeper** [comment]
Do the changes to fields/types/Type.js not address your concern?  

By setting `required` to a function we can control the validation.  The test file covers this unless I did something wrong.

**Comment by webteckie** [comment]
According to http://mongoosejs.com/docs/api.html#schematype_SchemaType-required it seems like @snowkeeper's  changes in Types.js should handle @JedWatson's concerns.  They seem to disable the mongoose required validation on that field type if evalDependsOn is falsy.

**Comment by ericelliott** [comment]
:+1:

**Comment by snowkeeper** [comment]
Here is the mongoose [issue](https://github.com/Automattic/mongoose/issues/941#issuecomment-61155319) that talks about using `required` as a function.  

The conversation above the linked comment starting [here](https://github.com/Automattic/mongoose/issues/941#issuecomment-38994781) covers our exact use case.

**Comment by snowkeeper** [comment]
We can actually extend this to so that the Admin UI respects  user supplied required functions in the schema.  

In `updateHandler.js` we can remove the `dependsOn` check and replace it with a function check
```
var isFunc = _.isFunction(field.required);
if (!isFunc || (isFunc && field.required.call(data))) {
	message = options.requiredMessages[field.path] || field.options.requiredMessage || field.label + ' is required';
	addValidationError(field.path, message);
	invalidated = true;	
}
```
Our `dependsOn` required function would then need to accept either a mongoose doc or plain Object.

I have these changes done if you want me to push them along.

**Comment by dodiaraculus17** [comment]
**Sorry about Advertising**

I need contributors for my project. If you would like to help the project is located here.
https://github.com/poKe-Websites/poKe-CMS

Thank You :)

**Comment by webteckie** [comment]
Can we get this merged if there's no other issues/concerns with this?

**Comment by JedWatson** [comment]
@snowkeeper and @webteckie thanks for clarifying - I didn't dig deep enough in my original review before asking the question. Looks like a great solution :grinning:

**Comment by webteckie** [comment]
@JedWatson thanks for merging @snowkeeper awesome work here!  @snowkeeper may want to add a bit of clarification to the docs about this.  @JedWatson you can close #1011 as a result of this merge.

**Pull request merged**